bluesky/protocols.py

```diff
@@ -191,7 +122,8 @@ def parent(self) -> Optional[Any]:
 class WritesExternalAssets(Protocol):
     @abstractmethod
     def collect_asset_docs(self) -> Iterator[Asset]:
```

Comment:
Make this `SyncOrAsync` too

Reply:
Actually, this doesn't quite work. `SyncOrAsync = Union[T, Awaitable[T]]`, but an `AsyncIterator` is not an `Awaitable[Iterator]`. I think we instead need:
```python
SyncOrAsyncIterator = Union[Iterator[T], AsyncIterator[T]]
...
    def collect_asset_docs(self) -> SyncOrAsyncIterator[Asset]:
```
The code that calls `collect_asset_docs` and friends will then need something different to `for doc in obj.collect_asset_docs()`. Something like:
```python
async def iterate_maybe_async(iterator: SyncOrAsyncIterator[T]) -> AsyncIterator[T]:
    if inspect.isasyncgen(ret):
        async for v in ret:
            yield v
    else:
        for v in ret:
            yield v
...
    async for doc in iterate_maybe_async(obj.collect_asset_docs()):
         ...
```
Similar for `collect` and `collect_pages`.